note to self so far i've done the bookmarks tabs from chrome for DGE and Bioinformtics

# RNA-seq analysis

## Differential gene expression

### Official vignettes by creators

* [RNA-seq workflow](http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html) R vignette, includes data exploration, DESeq, batch effects etc.
* [DESeq2 vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
* [txImport vignette](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html)
* [R vignette - salmon to DGE](https://www.bioconductor.org/packages/release/workflows/vignettes/rnaseqDTU/inst/doc/rnaseqDTU.html) by Mike Love et al.
* [edgeR user guide](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) (PDF)

### Other tutorials

* [RNA-seq tutorial with DESeq2](http://folk.uio.no/jonbra/MBV-INF4410_2017/exercises/2017-12-07_R_DESeq2_exercises_with_results.html) an online tutorial
* [Differential expression tutorial in EdgeR](http://www.compbio.dundee.ac.uk/user/pschofield/Projects/teaching_pg/workshops/biocDGE.html) By Piet√† Schofield.
* [RNASeq tutorial - preprocessing data](https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/02_Preprocessing_Data.nb.html) Cruk suummer schoolschool
* [DGE with limma-voom](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html) course material from UC davis.

## Brownian motion and OU modelling

* [Various online course content](http://www.eve.ucdavis.edu/~wainwrightlab/Roi/Site/Teaching.html) BY Sam Price and Roi Holzman, UC Davis. Contains OUCH tutorial.
* [evee tools](https://evee-tools.github.io/) scripts associated with Chen et al. 2019 OU paper


# Other transcriptomics

* [Grouper: Clustering contigs](https://github.com/COMBINE-lab/grouper)


# General coding resources

* [Search engine for symbols](http://symbolhound.com/)
* [regex tester 1](https://regex101.com/) and [regex tester 2](https://www.regextester.com/)
* [markkdown generator](https://dillinger.io/)
* [markdown table generator](https://www.tablesgenerator.com/markdown_tables)
* [GUI to help change your bashrc profile](http://ezprompt.net/)
* [Tao Te Programming](https://www.burns-stat.com/documents/books/tao-te-programming/table-of-contents/) Patrick Burns

## Bash

* [Explain Shell](https://explainshell.com/)
* [Shell Check](https://www.shellcheck.net/)
* various bash scripting guides:
    + [Handy bash one-liners](https://github.com/onceupon/Bash-Oneliner/blob/master/README.md)
    + [how to bash](http://tldp.org/HOWTO/Bash-Prog-Intro-HOWTO.html#toc8)
    + [bash templates](https://www.networkworld.com/article/2694348/unix--scripting-with-templates.html)
    + [writing robust bash scrippts](https://www.davidpashley.com/articles/writing-robust-shell-scripts/)
    + [advanced bash scripting guide](http://www.tldp.org/LDP/abs/html/index.html)
    + [coding style guide](https://github.com/robbyrussell/oh-my-zsh/wiki/Coding-style-guide)

## R

* [R for Data Science](https://r4ds.had.co.nz/) G Grolemund and H Wickham
* [R inferno book](https://www.burns-stat.com/pages/Tutor/R_inferno.pdf) Patric Burns 2011
* [Impatient R book](https://www.burns-stat.com/documents/tutorials/impatient-r/)
* Data viz in R [lecture](http://datacarpentry.org/semester-biology/materials/ggplot/) and [exercises](http://datacarpentry.org/semester-biology/assignments/r-datavis/) and [a textbook](https://r4ds.had.co.nz/data-visualisation.html)

## Bioinformatics

* [Handy bash commands for fasta files](https://www.biostars.org/p/17680/)
* [blastP](https://blast-ncbi-nlm-nih-gov.docelec.univ-lyon1.fr/Blast.cgi?PROGRAM=blastp&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome)

## Statistics

* Online course [Improving your statistical inferences](https://www.coursera.org/learn/statistical-inferences)

## Maths

* [Equation solver](https://www.symbolab.com/solver/equation-calculator/) takes an equation, solves it, and shows you the steps
